pkg/agent/controller/l7visibility/l7visibility_controller.go

```diff
@@ -0,0 +1,412 @@
+		if !ok {
+			nonAnnotatedPods = append(nonAnnotatedPods, pod)
+		}
+	}
+	return nonAnnotatedPods
+}
+
+func (l7c *L7VisibilityController) syncPod(podNN string) error {
+	podNamespace, podName := k8s.SplitNamespacedName(podNN)
+	pod, err := l7c.podLister.Pods(podNamespace).Get(podName)
+	if err != nil {
+		// Remove the TC flows if the Pod has been deleted
+		return l7c.removeTCFlow(podNN)
+	}
+	podNS, err := l7c.namespaceLister.Get(pod.Namespace)
+	if err != nil {
+		// Remove TC flows if Namespace has been deleted
+		return l7c.removeTCFlow(podNN)
+	}
+
+	var annotationValue string
+	podAnnotationValue, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+	if !ok {
+		// Both Pod and Namespace are not annotated, remove the TC Mark flow
+		nsAnnotationValue, ok := podNS.Annotations[types.L7VisibilityAnnotationKey]
+		if !ok {
+			return l7c.removeTCFlow(podNN)
+		}
+		annotationValue = nsAnnotationValue
+	}
+	// Pod has annotation, we will add it to the TCmarkflow
+	if annotationValue == "" {
+		annotationValue = podAnnotationValue
+	}
+
+	// Check if the annotation value is one of the specified values
+	var direction v1alpha2.Direction
+	annotationValue = strings.ToLower(annotationValue)
+	switch annotationValue {
+	case "ingress":
+		direction = v1alpha2.DirectionIngress
+	case "egress":
+		direction = v1alpha2.DirectionEgress
+	case "both":
+		direction = v1alpha2.DirectionBoth
+	default:
+		return errInvalidAnnotation
+	}
+	podInterfaces := l7c.interfaceStore.GetContainerInterfacesByPod(pod.Name, pod.Namespace)
+	if len(podInterfaces) == 0 {
+		return errPodInterfaceNotFound
+	}
+	sourceOfPort := []uint32{uint32(podInterfaces[0].OFPort)}
+
+	// Start Suricata before starting traffic control mark flows
+	l7c.l7Reconciler.StartSuricataOnce()
+
+	// Return if Pod with direction is already present in map
+	if l7c.ifPodNNAndDirectionExistsInMap(podNN, direction) {
+		return nil
+	}
+	// Remove TC mark flows if PodNN is present in map with different direction
+	l7c.removeTCFlow(podNN)
+
+	tcName := fmt.Sprintf("tcl7:%s", podNN)
+	if err := l7c.ofClient.InstallTrafficControlMarkFlows(tcName, sourceOfPort, l7c.targetPort, direction, v1alpha2.ActionMirror); err != nil {
+		return err
+	}
+	l7c.addPodNNDirToMap(podNN, direction)
+	return nil
+}
+
+func (l7c *L7VisibilityController) addPodNNDirToMap(podNN string, direction v1alpha2.Direction) {
+	l7c.podToDirectionMapMutex.Lock()
+	defer l7c.podToDirectionMapMutex.Unlock()
+	l7c.podToDirectionMap[podNN] = direction
+}
+
+func (l7c *L7VisibilityController) deletePodNNFromMap(podNN string) {
+	l7c.podToDirectionMapMutex.Lock()
+	defer l7c.podToDirectionMapMutex.Unlock()
+	delete(l7c.podToDirectionMap, podNN)
+}
+
+func (l7c *L7VisibilityController) getTCNameFromPodNNDirMap(podNN string) string {
+	l7c.podToDirectionMapMutex.RLock()
+	defer l7c.podToDirectionMapMutex.RUnlock()
+	if _, ok := l7c.podToDirectionMap[podNN]; ok {
+		return fmt.Sprintf("tcl7:%s", podNN)
+	}
+	return ""
+}
```

Comment:
It's redundant with `ifPodNNAndDirectionExistsInMap` (which could be renamed to `getMirroredDirection`), its caller should just call the latter and construct the tc name via the same function which generates the name for installation so the names will always be consistent and no redundancy. For example:
```
func (l7c *L7VisibilityController) removeTCFlow(podNN string) error {
	if _, exists := l7c.getMirroredDirection(); !exists {
		return nil
	}
	if err := l7c.ofClient.UninstallTrafficControlMarkFlows(generateTCName(podNN)); err != nil {
		return err
	}
	return nil
}

...
if err := l7c.ofClient.InstallTrafficControlMarkFlows(generateTCName(podNN), sourceOfPort, l7c.targetPort, direction, v1alpha2.ActionMirror); err != nil {
	return err
}
```